README
CGI-CRUD version 1.05
=====================

This module provide a framework to create web forms for a front-end
CRUD interface to a variety of data sources. It features
auto-discovery of the data dictionary for an RDBMS source.

With this framework, you can get a basic CRUD web interface up and
running in minutes without coding. At the same, it allows a great
deal of flexibility for customization by engineers (application
functionality/business logic) and non-engineers (presentation and
style) alike.

For an implementation see Cruddy!
http://www.thesmbexchange.com/cruddy/index.html

CRUD now, code later.

INSTALLATION

To install this module type the following:

   perl Makefile.PL
   make
   make test
   make install

DEPENDENCIES

This module requires these other modules and libraries:

  CGI::Enurl
  CGI::AutoForm
  DBIx::IO
  CGI::FastTemplate
  DBI
  
  Optional:
    If you plan to run this under plain ole' CGI then you'll need
      CGI
    
    To take full advantage of debugging with mod_perl_debug.pl (see Cruddy!
    referenced above)
      Data::Dumper

COPYRIGHT AND LICENCE

Copyright (C) 2000-2008 Reed Sandberg

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.



Changes
Revision history for Perl extension CGI::CRUD.

1.05  Tue Jul  8 13:57:54 PDT 2008
        - Release cleanup, documentation.

1.03  Tue Nov 27 03:33:42 2007
	- original version; created by h2xs 1.23 with options
		-A -X -b 5.5.3 -n CGI::CRUD



lib/CGI/CRUD.pm
package CGI::CRUD;

=head1 NAME

CGI::CRUD - Instant CRUD web front-end, featuring auto-discovery of the data dictionary for an RDBMS source

=head1 DESCRIPTION

This module provide a framework to create web forms for a front-end CRUD interface to a variety of data sources. It features auto-discovery of the data dictionary for an RDBMS source.

With this framework, you can get a basic CRUD web interface up and running in minutes without coding. At the same, it allows a great deal of flexibility
for customization by engineers (application functionality/business logic) and non-engineers (presentation and style) alike.

CRUD now, code later.


CRUD (Create, Read/Report, Update, Delete) are the four basic data manipulation commands of a data source; e.g. enabling management of configuration data in an administrative interface
to your application.

A unique and powerful advantage with this CRUD abstraction is that it can be tied closely with a database schema. Each group of fields in the form can represent a database table (or view) and the table/column properties and constraints are automagically discovered so your DBA can make DDL changes that will be immediately reflected in the HTML forms  - no duplication of the data dictionary in your code.

All user/operator input is checked tightly against database constraints and there is built-in magic to provide convenient select lists, etc, and to enforce a discreet set of valid values against unique/primary keys in lookup tables. This means referential integrity even for MySQL. Metadata in MySQL's C<SET> and C<ENUM> types are also supported.  This also gives the operator a chance to correct mistakes with helpful hints instead of just getting a meaningless db error code.

Another advantage this abstraction provides is the separation of presentation and style using style sheets and having human-friendly presentation attributes stored in a database table that can be managed by non-engineers.

=head1 SEE ALSO

For a quick start, see Cruddy!

L<http://www.thesmbexchange.com/cruddy/index.html>

For internals, start with L<CGI::CRUD::TableIO> and L<CGI::AutoForm>.

=head1 AUTHOR

Reed Sandberg, E<lt>reed_sandberg Ӓ yahooE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2000-2008 Reed Sandberg

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.


=cut

*CGI::CRUD::VERSION = \'1.05';

1;


lib/CGI/CRUD/Skipper.pm
#
# $Id: Skipper.pm,v 1.3 2005/01/27 21:33:26 rsandberg Exp $
#

package CGI::CRUD::Skipper;

use strict;
use CGI::CRUD::SkipperOutput;
use CGI::CRUD::TableIO;
use CGI::AutoForm;
use DBIx::IO::Table;

use constant OK => 0;

@CGI::CRUD::Skipper::ISA = (qw(CGI::CRUD::TableIO));

my $table_field = {
    FIELD_NAME => 'TABLE_NAME',
    INPUT_CONTROL_TYPE => 'SELECT',
    HEADING => 'Table Name',
    DATATYPE => 'CHAR',
    REQUIRED => 'Y',
    INSERTABLE => 'Y',
};

my $action_field = {
    FIELD_NAME => 'ACTION',
    INPUT_CONTROL_TYPE => 'RADIO',
    DEFAULT_VALUE => 'IR',
    HEADING => 'Operation',
    DATATYPE => 'CHAR',
    REQUIRED => 'Y',
    INSERTABLE => 'Y',
};
my $action_picklist = [
    { ID => 'IR', MASK => 'Insert', },
    { ID => 'SR', MASK => 'Search/Edit', },
];

my $table_picklist = undef();

=pod

=head1 NAME

CGI::CRUD::Skipper - Generic implementation of CGI::CRUD::TableIO

=head1 DESCRIPTION

A concrete subclass of CGI::CRUD::TableIO to provide a vanilla web front-end to your RDBMS. It performs auto-discovery of your schema
and data dictionary, just plug it in and go - you can customize and code later.

Naming conventions that bear any similarity to the famous TV star Muddy Mudskipper (having made several special appearances on "The Ren and Stimpy Show") are purely co-incidental.

=cut


sub handler
{
    my $r = new CGI::CRUD::SkipperOutput(@_) or return OK;

    my $self = __PACKAGE__->new($r->dbh());

    $self->{action} = $r->{apache}->uri() if $ENV{MOD_PERL};
    
    my $query = $r->query();
    #$query->{'__SDAT_TAB_ACTION.TABLE_NAME'} = uc($query->{'__SDAT_TAB_ACTION.TABLE_NAME'}) if exists($query->{'__SDAT_TAB_ACTION.TABLE_NAME'});
    my $action = $r->param('__SDAT_TAB_ACTION.ACTION') || '';

    # Set other vars to be substituted into CGI::FastTemplate
    $r->{tpl_vars}{HOME_URL} = $self->{action};

    if ($r->param('__SDAT_TAB_ACTION.RESTART'))
    {
        $self->request_action($r);
    }
    elsif ($action eq 'IR')
    {
        $self->insert_req($r);
    }
    elsif ($action eq 'ID')
    {
        $self->insert_data($r);
    }
    elsif ($action eq 'SR')
    {
        $self->search_req($r);
    }
    elsif ($action eq 'SD')
    {
        $self->search_results($r);
    }
    elsif ($action eq 'UR')
    {
        $self->update_req($r);
    }
    elsif ($action eq 'UD')
    {
        $self->update_data($r);
    }
    elsif ($action eq 'DR')
    {
        $self->delete_req($r);
    }
    else
    {
        $self->request_action($r);
    }
    return OK;
}

sub request_action
{
    my ($self,$r) = @_;
    my $dbh = $r->dbh();
    my $form = $r->form($dbh);
    $form->heading("Database Operations");
    $form->action($self->{action});
    $form->submit_value('Continue');
    $form->add_group('INSERTABLE',undef,'Choose a table and an operation to perform on that table','__SDAT_TAB_ACTION');

    unless (defined($table_picklist))
    {
        my $table_names = DBIx::IO::Table->existing_table_names($dbh);
        ref($table_names) or ($r->server_error(),return undef);
        my @tp = ();
        foreach my $table_name (@$table_names)
        {
            push(@tp,{ID => $table_name, MASK => $table_name});
        }
        $tp[0] = {ID => 'NO TABLES FOUND', MASK => 'NO TABLES FOUND'} unless @tp;
        $table_picklist = \@tp;
    }

    my $rv;
    unless ($rv = $form->add_field($table_field,$table_picklist))
    {
        defined($rv) or ($r->server_error(),return undef);
        $r->perror("Table list unavailable");
        return $rv;
    }
    $form->add_field($action_field,$action_picklist);
    $r->output($form->prepare($r->query()));
}

1;

__END__

=head1 BUGS

No known bugs.

=head1 SEE ALSO

L<CGI::CRUD::TableIO>

=head1 AUTHOR

Reed Sandberg, E<lt>reed_sandberg Ӓ yahooE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2000-2007 Reed Sandberg

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.



lib/CGI/CRUD/ApacheOutputFastTemplate.pm
#
# $Id: ApacheOutputFastTemplate.pm,v 1.5 2005/01/27 23:30:50 scottb Exp $
#

package CGI::CRUD::ApacheOutputFastTemplate;

use strict;
use CGI::AutoForm;
use CGI::FastTemplate;
use CGI::CRUD::Output;

@CGI::CRUD::ApacheOutputFastTemplate::ISA = qw(CGI::CRUD::Output);


# CGI::FastTemplate path for templates
# May override by setting CRUDDY_FAST_TEMPLATE_PATH env
# variable (e.g. SetEnv or PerlSetEnv)
my $DEFAULT_TPL_PATH = '/var/www/tpl';

# CGI::FastTemplate 'main' template
# May override by setting CRUDDY_FAST_TEMPLATE_MAIN env
# variable (e.g. SetEnv or PerlSetEnv)
# -or-
# as a second argument to the C<output> method call
my $DEFAULT_TPL_MAIN = 'cruddy.tpl';

#
# Use in conjunction with mod_perl_debug.pl
# NOTE!
# Do not use in production sites
#
my $PARAMDUMP = $ENV{CRUDDY_PARADUMP_DEBUG};


if ($PARAMDUMP)
{
    # To convert carps to clucks ...
    #CGI::LogCarp::import( 'verbose' );
}


sub new
{
    my $caller = shift;
    my ($r,$defaults) = @_;

    my $self = $caller->SUPER::new(@_);

    if ($ENV{MOD_PERL})
    {
        $self->{apache} = $r;
        $self->{user} = $r->user();

        # Save the Apache Request instance so its methods can be
        # accessed later.
        $self->{apreq} = $::MOD_PERL_REQ_CLASS->new($r);

        my @pars = $self->{apreq}->param();
        my %query;
        foreach my $par (@pars)
        {
            my @t = $self->{apreq}->param($par);
            if (@t > 1)
            {
                $query{uc($par)} = \@t;
            }
            else
            {
                $query{uc($par)} = $t[0];
            }
        }
        $self->{q} = \%query;

        if ($ENV{CRUDDY_PARADUMP_DEBUG})
        {
            require "Data/Dumper.pm";
            open(PDUMP,">$ENV{CRUDDY_PARADUMP_DEBUG}") || die($!);
            print PDUMP Data::Dumper::Dumper(\%query);
            print PDUMP "1;\n";
            close(PDUMP);
        }
    }
    else
    {
        $self->{user} = $r->[1];
        my %t = ();
        if (ref($::VAR1) eq 'HASH')
        {
            %t = %$::VAR1;
        }
        my $i = 0;
        my %q;
        while (defined(${$r->[0]}[$i]))
        {
            my $key = uc(${$r->[0]}[$i]);
            my $val = ${$r->[0]}[++$i];
            push(@{$q{$key}},$val);
            $i++;
        }
        my ($key,$val);
        while (($key,$val) = each(%q))
        {
            $q{$key} = $val->[0] unless $#$val;
        }
        $self->{q} = { %t,%q };
    }

    $self->{tpl_vars} = $defaults;

    return $self;
}

sub form_attrs
{
    my ($caller,$form) = @_;
    $caller->SUPER::form_attrs($form);
    $form->{GT} = qq[WIDTH="80%" CELLPADDING="5" CELLSPACING="0" BORDER="0"];
    $form->{VFL} = qq[WIDTH="40%" ALIGN="RIGHT"];
    return $form;
}

##at should somehow cache templates
##at Accept scalar references
sub output
{
    my ($self,$out,$tplf) = @_;
    $tplf = $ENV{CRUDDY_FAST_TEMPLATE_MAIN} || $DEFAULT_TPL_MAIN unless $tplf;

    my $tpl = new CGI::FastTemplate($ENV{CRUDDY_FAST_TEMPLATE_PATH} || $DEFAULT_TPL_PATH);
##at should check return value because an OS call was done
    $tpl->define(main => $tplf);
    $self->{tpl_vars}{BODY} = (ref($out) ? $$out : $out) if defined($out);
    $tpl->assign($self->{tpl_vars});
    $tpl->parse(CONTENT => 'main');
    my $gob = $tpl->fetch('CONTENT');

    $self->{apache}->no_cache(1) if $ENV{MOD_PERL};
    $self->{apache}->content_type('text/html') if $ENV{MOD_PERL};
    $self->{apache}->print($$gob) if $ENV{MOD_PERL};
    return 1;
}

1;


lib/CGI/CRUD/TableIO.pm
#
# $Id: TableIO.pm,v 1.3 2005/01/27 21:33:26 rsandberg Exp $
#


package CGI::CRUD::TableIO;

use strict;
use DBIx::IO::Search;
use DBIx::IO::Table;
use CGI::CRUD::Table;
use CGI::AutoForm;
use CGI::Enurl ();

use constant OK => 0;

=pod

=head1 NAME

CGI::CRUD::TableIO - Virtual base class for a basic web front-end to an RDBMS

=head1 DESCRIPTION

Virtual base class provides skeletal CRUD routines for a web front-end. Subclass/override/customize
to your heart's content. One popular method to override is C<verify_input>.

=cut

# ctor
sub new
{
    my ($caller,$dbh) = @_;

    my $class = ref($caller) || $caller;
    my $obj = { dbh => $dbh };

    $obj = bless($obj,$class);

    $obj->{verify_input} = sub { $obj->verify_input(@_); };
    return $obj;
}

# Build SQL for the search form submission
sub where_sql
{
    my ($self,$table_dat,$table_name) = @_;
    
    my $searcher = $self->build_search($table_dat,$table_name) or return undef;
    $searcher->_build_sql();
    return $searcher->{where} || 0;
}

# Build the DBIx::IO::Search object to assist with where_sql()
sub build_search
{
    my ($self,$table_dat,$table_name) = @_;
    
    my $searcher = new DBIx::IO::Search($self->{dbh},$table_name) or return undef;
    my ($field,$val);
    while (($field,$val) = each(%$table_dat))
    {
        next if !length($val);
        if (ref($val) eq 'ARRAY')
        {
            $searcher->build_list_crit($field,$val);
        }
        elsif (ref($val) eq 'HASH')
        {
            if (exists($val->{_WM}))
            {
                $searcher->build_scalar_crit($field,'LIKE',$val->{_WM});
            }
            else
            {
                # Expect 2 keys from this hash for date range values
                $searcher->build_range_crit($field,$val->{_RS},$val->{_RE}) 
                    if (($searcher->{io}->is_date($field) || $searcher->{io}->is_datetime($field)) ? $val->{_UR} : 1);
            }
        }
        else
        {
            $searcher->build_scalar_crit($field,'=',$val);
        }
    }
    
    return $searcher;
}

# Perform record deletion operation
sub delete_req
{
    my ($self,$r) = @_;
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    my $q = $r->query();
    my ($table,$rec);
    my $sq = CGI::AutoForm->format_query($q);
    unless (($table = new DBIx::IO::Table($r->dbh(),undef,undef,$table_name)) && ($rec = $table->fetch($sq->{__SDAT}{KEYS})))
    {
        defined($rec) && ($r->output("Record no longer exists"),return OK);
        $r->server_error();
        return undef;
    }
    unless ($table->delete())
    {
        $r->server_error();
        return undef;
    }
    my $msg = qq[<P>Record Deleted</P>];
    $msg .= $self->return_results($q);
    $r->output($msg);
}

sub return_results
{
    my ($self,$fq) = @_;
    my $msg = qq[<P><TABLE WIDTH="100%"><TR>];
    my $q = CGI::AutoForm->extract_query_group($fq,'__SDAT_TAB_ACTION');
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'SD';
    my $stq = CGI::AutoForm->extract_cut_query_group($fq,'__SDAT.SC');
    my $sq = stringify_query({ %$q, %$stq });
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">Return to search results<A></TD>];
    $msg .= $self->sreturn($q);
    $msg .= qq[</TR></TABLE></P>];
    return $msg;
}

sub sreturn
{
    my ($self,$q) = @_;
    my $msg;
    my $eq = CGI::AutoForm->extract_query_group($q,'__SDAT_TAB_ACTION');
    $eq->{'__SDAT_TAB_ACTION.ACTION'} = 'SR';
    my $sq = stringify_query($eq);
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">New Search with $q->{'__SDAT_TAB_ACTION.TABLE_NAME'}<A></TD>];
    $eq->{'__SDAT_TAB_ACTION.ACTION'} = 'IR';
    $sq = stringify_query($eq);
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">Add to $q->{'__SDAT_TAB_ACTION.TABLE_NAME'}<A></TD>];
    $eq->{'__SDAT_TAB_ACTION.RESTART'} = 1;
    $sq = stringify_query($eq);
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">New DB Operation<A></TD>];
    return $msg;
}

sub new_start
{
    my ($self,$q) = @_;
    my $msg = qq[<P><TABLE WIDTH="100%"><TR>];
    $msg .= $self->sreturn($q);
    $msg .= qq[</TR></TABLE></P>];
    return $msg;
}

# Perform update operation

# special value of NULL still recognized, however its sufficient to have an empty new value
# where the existing value is not empty, this will update the value to NULL a little more risky but much more
# convenient because values of length < 4 (e.g. YORN and date elements) will have to be expanded to 4
# losing some ability to constrain the values
# THIS MEANS IT IS UP TO YOU TO REPRESENT ALL VALUES IN AN UPDATE, OTHERWISE THEY **WILL BE SET TO NULL**
# e.g. submit a full record to form->add_record and make sure field_template has *all* fields, either by
# completely relying on the data dictionary or inserting a record for all fields in UI_TABLE_COLUMN
sub update_data
{
    my ($self,$r) = @_;
    my $form = $self->update_form($r) || return undef;
    my $q = $r->query();
    my %vq = %$q;
    map { $vq{$_} =~ s/^NULL$// } keys(%vq);
    unless ($form->validate_query(\%vq,$self->{verify_input}))
    {
        $r->output($form->prepare($q));
        return OK;
    }

    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    my ($table,$rec);
    my $sq = $form->format_query($q);
    unless (($table = new CGI::CRUD::Table($r->dbh(),$r->user,undef,undef,$table_name)) && ($rec = $table->fetch($sq->{__SDAT}{KEYS})))
    {
        defined($rec) && ($r->output("Record no longer exists"),return OK);
        $r->server_error();
        return undef;
    }
    my $table_dat = $sq->{uc($table_name)};

    # a special value of 'NULL' updates a value to NULL
    map { $table_dat->{$_} =~ s/^NULL$// } keys(%$table_dat);
    map { $table_dat->{$_} = '' unless exists($table_dat->{$_}) } keys(%{$table->column_types()});
    unless ($table->update($table_dat))
    {
        $r->server_error();
        return undef;
    }
    my $msg = qq[<P>Record Updated</P>];
    $msg .= $self->return_results($q);
    $r->output($msg);
}

# Build update form
sub update_form
{
    my ($self,$r) = @_;
    my $form = $r->form($r->dbh());
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    $form->heading("Update $table_name");
    $form->action($self->{action});
    $form->submit_value('Update');
    $r->graceful_add_form_group($form,'DISPLAY EDIT',$table_name,'Edit fields and submit when done') || return undef;
    return $form;
}

# Build/present update form
sub update_req
{
    my ($self,$r) = @_;
    my $form = $self->update_form($r) || return undef;
    my $q = $r->query();
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'UD';
    my $sq = $form->format_query($q);
    my ($table,$rec);
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    unless (($table = new DBIx::IO::Table($r->dbh(),undef,undef,$table_name)) && ($rec = $table->fetch($sq->{__SDAT}{KEYS})))
    {
        defined($rec) && ($r->output("Record no longer exists"),return OK);
        $r->server_error();
        return undef;
    }
    $form->add_record($rec);
    $r->output($form->prepare($q));
}

# Perform search operation and return results
sub search_results
{
    my ($self,$r) = @_;
    
    # keep in mind this is NOT normalized or unescaped
    my $q = $r->query();
    my $form = $self->search_form($r) || return undef;
    unless ($form->validate_query($q))
    {
        $r->output($form->prepare($q));
        return OK;
    }

    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    $form = new CGI::AutoForm($r->dbh());
    $r->graceful_add_form_group($form,'DISPLAY ONLY',$table_name,"Searching...",undef,1) || return undef;
    $q = $form->format_query($r->query());
    my $table_dat = $q->{uc($table_name)};
    my $searcher = $self->build_search($table_dat,$table_name) or ($r->server_error(),return undef);
    my $field_list = $form->field_list();
    my $ffield;
    foreach my $f (@$field_list)
    {
        if (length($f->{BRIEF_HEADING}))
        {
            $ffield = $f->{FIELD_NAME};
            last;
        }
    }
    $searcher->sortlist([ $ffield ]);

    my $results = $searcher->search();
    unless ($results)
    {
        $r->server_error();
        return $results;
    }
    unless (@$results)
    {
        $r->output("No results found");
        return 1;
    }
    my $recno = scalar(@$results);
    my $recapp = ($recno > 1 ? 's' : '');

    my $keys = $searcher->{pk};
    my $warnk = '';
    unless (@$keys)
    {
##at could workaround this for Oracle by using ROWID
        #warn("search requested for a table with no primary key");
        $warnk = qq[<P><FONT COLOR="RED">View/Edit/Delete of individual records disabled because there is no primary key defined for this table.</FONT></P>];
    }

    $form->current_group()->{heading} = "Found $recno record$recapp.$warnk";
    $form->add_record($results);

    # requires that $table_name has a primary key (via DBIx::IO)

    my $tq = $form->extract_query_group($r->query(),uc($table_name));
    my $search_cache = stringify_query($tq,"__SDAT.SC");

    # make sure there are no closure issues with, for example, $searcher shouldn't be any with an anonymous sub
    my $rec_callback = sub
    {
        my ($rec_html,$rec,$group) = @_;
        my $add;
        my $etab = CGI::Enurl::enurl($table_name);
        my $qs = CGI::Enurl::enURL(qq[__SDAT_TAB_ACTION.TABLE_NAME=]) . CGI::Enurl::enurl($table_name) . '&';
        my $ur = CGI::Enurl::enURL(qq[__SDAT_TAB_ACTION.ACTION=UR&]);
        my $dr = CGI::Enurl::enURL(qq[__SDAT_TAB_ACTION.ACTION=DR&]);
        foreach my $key (@$keys)
        {
            $qs .= CGI::Enurl::enurl("__SDAT.KEYS.$key") . '=' . CGI::Enurl::enurl($rec->{$key}) . "&";
        }
        chop($qs);
        $add .= qq[<TD><A HREF="$self->{action}?$ur$qs&$search_cache">View/Edit</A></TD>];
        $add .= qq[<TD style="width: 10px; text-align: center;">|</TD><TD><A HREF="$self->{action}?$dr$qs&$search_cache">Delete</A></TD>];
        return $rec_html . $add;
    };
    $form->{head_html} = qq[<H2>Search Results for $table_name</H2>];

    $form->{tail_html} = ' ';
    my $html = $form->prepare(undef,undef,(@$keys ? $rec_callback : undef()));
    $$html .= $self->new_start($r->query());
    $r->output($html);
}

# Build search form
sub search_form
{
    my ($self,$r) = @_;
    my $form = $r->form($r->dbh());
    $form->heading('Search Criteria');
    $form->action($self->{action});
    $form->submit_value('Search');
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    $r->graceful_add_form_group($form,'SEARCHABLE',$table_name,"Build criteria to report on $table_name") || return undef;
    return $form;
}

# Build/present search form
sub search_req
{
    my ($self,$r) = @_;
    my $form = $self->search_form($r) || return undef;
    my $q = $r->query();
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'SD';
    $r->output($form->prepare($q));
}

# Override in subclass to perform custom checks on input
# Will be passed as a callback to CGI::AutoForm::validate_query()
# refer to those docs for parameters and expected return values
sub verify_input
{
    return 1;
}

# Perform insert operation
sub insert_data
{
    my ($self,$r) = @_;
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    my $q = $r->query();
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'ID';
    my $form = $self->insert_form($r->dbh(),$table_name,$r) || return undef;
    unless ($form->validate_query($q,$self->{verify_input}))
    {
        $r->output($form->prepare($q));
        return OK;
    }

    my $table;
    my $sq = $form->format_query($q);
    my $rv;
    unless (($table = new CGI::CRUD::Table($r->dbh(),$r->user,undef,undef,$table_name)) && ($rv = $table->insert($sq->{uc($table_name)})))
    {
        $r->server_error();
        return undef;
    }
    if ($rv == -1.1)
    {
        $r->perror("No data to insert");
        return undef;
    }
    elsif ($rv == -1.4)
    {
        $r->perror("Duplicate key violation on insert.");
        return undef;
    }
    my $aq = $form->extract_query_group($q,'__SDAT_TAB_ACTION');
    if ($q->{'__SDAT.CONTINUE'})
    {
        $form->{top_message} = qq[<DIV>Data saved.</DIV>];
        $r->output($form->prepare($aq));
    }
    else
    {
        $r->output($self->insert_or_return($aq));
    }
}

# Build insert form
sub insert_form
{
    my ($self,$dbh,$table_name,$r) = @_;
    my $form = $r->form($dbh);
    $form->heading("Input for $table_name");
    $form->action($self->{action});

    $form->{tail_html} = qq[<P><TABLE WIDTH="100%"><TR ALIGN="CENTER"><TD><INPUT TYPE="RESET">&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;] .
        qq[<INPUT TYPE="SUBMIT" NAME="__SDAT.CONTINUE" VALUE="Save/Continue">] .
        qq[&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;<INPUT TYPE="SUBMIT" VALUE="Save"></TD></TR></TABLE></P></FORM>];
    $r->graceful_add_form_group($form,'INSERTABLE',$table_name,"Insert data for \U$table_name") || return undef;
    return $form;
}

sub insert_or_return
{
    my ($self,$q,$top_message) = @_;
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'IR';
    my $msg;
    $msg .= 'Data saved.';
    $msg .= "<BR>$top_message" if $top_message;
    $msg .= qq[<P><TABLE WIDTH="100%"><TR>];
    my $sq = stringify_query($q);
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">Insert new record<A></TD>];
    $q->{'__SDAT_TAB_ACTION.RESTART'} = 1;
    $sq = stringify_query($q);
    $msg .= qq[<TD><A HREF="$self->{action}?$sq">Admin panel<A></TD>];
    $msg .= qq[</TR></TABLE></P>];
    return $msg;
}

# Build/present insert form
sub insert_req
{
    my ($self,$r) = @_;
    my $table_name = $r->param('__SDAT_TAB_ACTION.TABLE_NAME');
    my $q = $r->query();
    $q->{'__SDAT_TAB_ACTION.ACTION'} = 'ID';
    my $form = $self->insert_form($r->dbh(),$table_name,$r) || return undef;
    $r->output($form->prepare($q));
}

sub stringify_query
{
    my ($q,$prefix) = @_;
    if (length($prefix))
    {
        $prefix .= '.' unless substr($prefix,-1) eq '.';
    }
    my ($str,$field,$val);
    my %dates_expand;
    foreach my $fq (@{flatten_query($q)})
    {
        ($field,$val) = each(%$fq);
        if ($field =~ /(.*)\._(RE\.|RS\.|UR$)/)
        {
            push(@{$dates_expand{$1}{params}},{$field,$val});
            $dates_expand{$1}{ur} = 1 if $2 eq 'UR' && length($val);
        }
        else
        {
            $str .= CGI::Enurl::enurl("$prefix$field") . '=' . CGI::Enurl::enurl($val) . '&' if length($val);
        }
    }

    my ($f,$v);
    while (($f,$v) = each(%dates_expand))
    {
        if ($dates_expand{$f}{ur})
        {
            foreach my $fq (@{$dates_expand{$f}{params}})
            {
                ($field,$val) = each(%$fq);
                $str .= CGI::Enurl::enurl("$prefix$field") . '=' . CGI::Enurl::enurl($val) . '&' if length($val);
            }
        }
    }
    chop($str);
    return $str;
}

# Do not pass in a structured query
sub flatten_query
{
    my ($q) = @_;
    my (@flat,$name,$val);
    while (($name,$val) = each(%$q))
    {
        $val = [ $val ] unless ref($val);
        foreach my $v (@$val)
        {
            push(@flat,{ $name => $v });
        }
    }
    return \@flat;
}

1;

__END__

=head1 SEE ALSO

L<CGI::AutoForm>, L<DBIx::IO>, L<DBIx::IO::Table>, L<CGI::CRUD::Table>

=head1 AUTHOR

Reed Sandberg, E<lt>reed_sandberg Ӓ yahooE<gt>

=head1 COPYRIGHT AND LICENSE

Copyright (C) 2000-2007 Reed Sandberg

This library is free software; you can redistribute it and/or modify
it under the same terms as Perl itself.

The full text of the license can be found in the
LICENSE file included with this module.

